### ISSUE TITLE
[Feature]: Simple Data Parallelism in vLLM

### ISSUE THREAD

**Issue by simon-mo** [vllm-project/vllm#9206]
### 🚀 The feature, motivation and pitch

It is common to have a scenario where folks want to deploy multiple vLLM instances on a single machine due to the machine have several GPUs (commonly 8 GPUs). The work can then be sharded across replicated instances. This issue describes the intended UX for such feature. Notably we might not want to tackle large distributed settings (100s of parallel vLLM instances), which should be better handled by a higher layers.

* Offline use case, for the LLM class, a new argument data_parallel_size and support dispatching requests to one engine per GPU (or per tensor parallel size).

```python
from vllm import LLM

llm = LLM(model="...", data_parallel_size=X) # spawn X number of engine processes and shard the work among them
llm = LLM(model="...", data_parallel_size=X, tensor_parallel_size=Y) # this is supported if X*Y <= total number of GPUs
```

For the server, same argument, route requests to different engine processes, we can start with simple round robin load balancing, but a good stretch goal is session affinity or prefix aware routing

```bash
vllm serve ... --data-parallel-size X
```

### Alternatives

* LiteLLM + manually creating replicas
* Using ray.data or ray serve to scale out

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by archthegit** [comment]
Hi, I'll be working on this issue

**Comment by BKitor** [comment]
I like this idea, providing flexibility around deployment is great!
Being able to deploy multiple types of parallelism would be ideal, e.g. `-pp 2 -dp 4 -dp 2` across 16 GPUs would be benificial. 
Would also be good to have fine-grained control of process-to-device topology mapping, but that might be a stretch goal for a different ticket...

**Comment by andoorve** [comment]
Related issue: https://github.com/vllm-project/vllm/issues/9198

**Comment by Imss27** [comment]
Previously, I encountered an issue: When running vLLM(older version) on a machine with multiple GPUs(a CI environment) without specifying `CUDA_VISIBLE_DEVICE`, it keeps allocating memory on one single GPU which then causes OOM error. This might be related?

@archthegit and I chated a little and plan to investigate and potentially work together on this.😄

**Comment by noooop** [comment]
Specifying CUDA_VISIBLE_DEVICE before starting vllm can use different cards for multiple vllm instances started on one machine. But multiple vllm instances may have port conflicts. So the best way to use multiple vllm instances on a single machine is to use docker

Now implementation of pp/tp is coupled with layers, it is very difficult to implement dp. If you don't use pp/tp, you can patch GroupCoordinator to implement dp, but without asynchronous scheduling, there is 
no meaning to implementing dp. Therefore, asynchronous scheduling should be implemented first before implementing dp

**Comment by noooop** [comment]
#8452 

A prototype by patching GroupCoordinator to implement dp.

**Comment by tensorflowt** [comment]
> #8452
> 
> A prototype, first implement asynchronous scheduling, then by patching GroupCoordinator to implement dp.

What is patching group coordinator ? Can I verify the performance of dp through benchmark_throughput.py?

**Comment by jiqiujia** [comment]
> Specifying CUDA_VISIBLE_DEVICE before starting vllm can use different cards for multiple vllm instances started on one machine. But multiple vllm instances may have port conflicts. So the best way to use multiple vllm instances on a single machine is to use docker
> 
> Now implementation of pp/tp is coupled with layers, it is very difficult to implement dp. If you don't use pp/tp, you can patch GroupCoordinator to implement dp, but without asynchronous scheduling, there is no meaning to implementing dp. Therefore, asynchronous scheduling should be implemented first before implementing dp
> 
> If you have many GPUs doing decoder-only model inference, using "Disaggregating Prefill and Decoding architecture", compare with use dp, has higher throughput , refer to [DistServe](https://arxiv.org/abs/2401.09670) [Mooncake](https://arxiv.org/abs/2407.00079)

It seems that sglang supports dp through lanuching multiple servers, e.g. https://github.com/sgl-project/sglang/blob/main/docs/backend/server_arguments.md, why don't vllm just implement like this?

**Comment by simon-mo** [comment]
Reopening this to focus on simple offline use case.

**Comment by rangehow** [comment]
any update on this? 👀

**Comment by izhuhaoran** [comment]
> Reopening this to focus on simple offline use case.

Is there any progress on this, or is vllm not currently planning to enhance this?

**Comment by clefourrier** [comment]
@simon-mo I need this for a project so I could take a look if no one's working on it - would it be useful?

**Comment by tianyilt** [comment]
爷青回了！就连vllm都愿意支持dp了,在0.8.5版本
Grandpa Qing returned! Even vllm is willing to support dp, in version 0.8.5
```
vllm serve /path_to/models--Qwen--Qwen2.5-VL-32B-Instruct/snapshots/7cfb30d71a1f4f49a57592323337a4a4727301da/ --dtype bfloat16 --port 8000 --host 0.0.0.0 --served-model-name qwen --trust-remote-code --gpu_memory_utilization 0.92 --seed 42 --tensor_parallel_size 2 --data-parallel-size 2
```

**Comment by vadimkantorov** [comment]
I have the same issue in https://github.com/vllm-project/vllm/issues/1237#issuecomment-2910167215

**Comment by vadimkantorov** [comment]
Does the `vllm serve --data-parallel-size 4` launch multiple `vllm serve` worker processes? (e.g. with `--tensor-parallel-size 2`, there should be 4 workers, each using 2 gpu's for sharding)

- Are these workers processing the requests independently?
- If any of these workers fail, will `vllm serve` survive gracefully and still process the prompts in other workers?
- Will vllm main process auto-restart failed worker processes? (e.g. because of OOMs)

Thanks! Trying to understand the existing support in vllm for using all GPUs for small models served on a 8-GPU machihe (e.g. model using only 1 GPU, or model using only 2 GPUs)

I think it's super important to have a simple out-of-the-box API serving which can reliably use/load-balance all resources of a single-node multi-gpu setup without resorting to more complex solutions requiring k8s etc (like https://github.com/vllm-project/production-stack)

And another usercase is super-important too: very simple offline processing of a dataset of inputs, which can use all of GPUs and spawn multiple engines on a single-node to process an offline pre-existing set of queries.

**Comment by yfzhou3993** [comment]
Can using vllm DP improve throughput? I deployed a 3B model with two different settings: TP=4 and TP=4 DP=2. However, the throughput of the former was slightly higher.

Could it be that DP enforces eager mode, which in turn reduces the throughput?

**Comment by DarkLight1337** [comment]
@njhill is this still not resolved?

**Comment by vadimkantorov** [comment]
Yes, apparently vllm serve with DP  enforces eager mode :( it is printed in the log even

**Comment by njhill** [comment]
I think this issue can be considered complete now. Data parallel deployment is documented [here](https://docs.vllm.ai/en/latest/serving/data_parallel_deployment.html). There is ongoing work on some improvements, in particular fault tolerance / failure recovery. For offline cases, we might consider also supporting at the `LLM` level rather than requiring an additional script to partition the data and launch the processes.

> apparently vllm serve with DP enforces eager mode :( it is printed in the log even

@vadimkantorov CUDA graphs should now also be supported with DP. I'll double check that this is the case.

**Comment by vadimkantorov** [comment]
> I'll double check that this is the case.

Yeah, as I got this in the log:
```
INFO 08-01 12:18:31 [config.py:1869] Defaulting to use mp for distributed inference
INFO 08-01 12:18:31 [config.py:2112] Chunked prefill is enabled with max_num_batched_tokens=8192.
INFO 08-01 12:18:31 [cuda.py:156] Data Parallel: Forcing enforce eager to be True since DP is currently not supported with CUDA Graphs.
```
And RunLLM tried to gaslight me somewhat about the reasons for this: https://discuss.vllm.ai/t/vllm-serve-to-use-all-gpus-of-a-single-node-multi-use/1264/ :( I'm feeling some strong uncanny valley vibes. Would be good to be able to ask RunLLM to abstain from responding (maybe via some special token/word in the thread), there is also no way to dislike its responses to provide publicly-visible feedback.

---

I also wonder if vllm caches Inductor outputs and autotune results between the runs to make torch.compile much faster (maybe in Mega-Cache: https://docs.pytorch.org/tutorials/recipes/torch_compile_caching_tutorial.html via using `torch.compiler.save_cache_artifacts()` and `torch.compiler.load_cache_artifacts()`. does vllm use tehse? there is even support for remote cache if vllm wants to provide some centralized compilation artifacts for the model it supports: https://docs.pytorch.org/tutorials/recipes/torch_compile_caching_configuration_tutorial.html, it might also be possible to extend it for loading artifacts from static website / github release). Measuring start-up time is also important - especially for DP when all ranks start to overload the same machine's I/O to load a lot of weights from disk to VRAM. Does DP set-up use nccl to transfer the loaded weights from one GPU to others? Or all ranks would load the weights from disk simultaneously?

**Comment by njhill** [comment]
@vadimkantorov it looks like you're using an out-dated version, this has been supported since 0.9.1: https://github.com/vllm-project/vllm/pull/18724.

Re torch compile artifact reuse, there is a workstream looking at improving startup time in general including this, see https://github.com/vllm-project/vllm/issues/19824 and the `#feat-startup-ux` channel in vLLM slack.

> Does DP set-up use nccl to transfer the loaded weights from one GPU to others? Or all ranks would load the weights from disk simultaneously?

This is a good point. I have noticed a startup slowdown due to apparent IO contention of the common layers, so propagating via NCCL in this case could be a good enhancement to look into - even better dividing the relevant layers between the DP ranks and do an A2A.

**Comment by vadimkantorov** [comment]
One last thing: in DP mode, are actually forward methods supposed to be invoked simultaneously and block/wait for other ranks? Or are the workers completely independent and require no sync at all (as if manually launched different engines with some manually request dispatching)?

The difference might be if differently-sized batches get allocated to different workers, and fast worker has to wait for the slow worker. Or if one of engines crashes / takes too long to restart / or crashes constantly / or one particular request-prompt always leads to crash no matter what

**Comment by njhill** [comment]
@vadimkantorov DP is usually used for MoE models where the expert layers use TP and/or EP. This requires forward passes to be synchronized, including processing empty batches when only a subset of the ranks have DP requests. This is handled automatically in the current implementation. W.r.t. keeping the ranks balanced, there is a built-in load balancer for the online serving mode (not for offline currently).

This synchronization isn't necessary for non-MoE models (where the DP would be completely independent) but it will currently still be done. It's designed to have negligible overhead but we could consider an improvement to disable it in those cases.

I'm not sure what you mean about crashing. Workers shouldn't crash and it implies a bug or misconfiguration. As mentioned we do still want to improve resiliency around this however.

**Comment by vadimkantorov** [comment]
> DP is usually used for MoE models where the expert layers use TP and/or EP

> This synchronization isn't necessary for non-MoE models (where the DP would be completely independent) but it will currently still be done. It's designed to have negligible overhead but we could consider an improvement to disable it in those cases.

I propose that the current functioning be explained in the docs: https://docs.vllm.ai/en/stable/configuration/engine_args.html#-data-parallel-size-dp , currently it's very concise, and does not explain the functioning for non-MoE models

**Comment by HackGiter** [comment]
Any progress on it? I think it would be quite useful for efficient inference on Qwen3-4B with a single node and 8 GPUs.

**Comment by vadimkantorov** [comment]
@njhill So maybe some sort of non-synchronized DP is needed, or some sort of built-in simple proxy/load-balancer - so that `vllm serve` could use all GPUs and that all GPUs would not need to wait for one GPU which happened to get a very long prompt (which probably would happen in the DP mode currently)...

I think it's a super-frequent usecase, and would be great to not need any extra things like Triton Inference Server or other wrappers for just using all GPUs of a single-node

**Comment by njhill** [comment]
@vadimkantorov yes actually I opened an issue for this https://github.com/vllm-project/vllm/issues/24461.

**Comment by njhill** [comment]
Here's a PR to address the unnecessary non-MoE overhead: https://github.com/vllm-project/vllm/pull/30739.

@vadimkantorov it would be great if you could open new issue(s) for the ideas you described above, thanks!
